### DISCUSSION TITLE
Welcome to python_course Discussions!

### DISCUSSION THREAD

**Discussion by olasupo** [olasupo/python_course#1] [Announcements]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by WaqarAbbas12** [comment]
I have been studying DSA and don't understand how ctypes.py_object works. As of my knowledge it works on concept of pointers to create an array of pointers that will store memory addresses for values to be added in dynamic array.

I have provided my code below:

def __create_array(self, space):
    return (space * ctypes.py_object)()
Basically, this function will create an array of pointer objects with memory addresses to store values in array. right?

**Reply by suppy376** [reply]
Hi Waqar, 

ctypes.py_object is a type provided by Python's ctypes module, which is used for low-level memory manipulation and interfacing with C libraries. The ctypes module allows you to create, manipulate, and pass C-compatible data types to functions written in C or to manage memory manually.

**Comment by suppy376** [comment]
So, Your understanding is correct.

The `__create_array` function creates an array that can store references (pointers, in a sense) to Python objects. These references point to the actual memory locations where the objects are stored. This is why it's dynamic—you're not limited to a specific type or size, and you can store any Python object in these slots.

Please, let me know if you need further clarification!

**Reply by WaqarAbbas12** [reply]
Thanks alot dear sir, So ctypes makes an array of pointers  containing memory addresses for values to store. So by default this array could be considered as an array of null pointers containing memory address. right?

**Reply by olasupo** [reply]
The contents of the array will be undefined (or garbage). They won't automatically be NULL pointers unless you explicitly initialize them to NULL or zero.
